deepchem/rl/mcts.py

```diff
@@ -0,0 +1,411 @@
+                    state,
+                    deterministic=False):
+    """Select an action to perform based on the environment's state.
+
+    Parameters
+    ----------
+    state: array
+      the state of the environment for which to select an action
+    deterministic: bool
+      if True, always return the best action (that is, the one with highest probability).
+      If False, randomly select an action based on the computed probabilities.
+
+    Returns
+    -------
+    the index of the selected action
+    """
+    if not self._state_is_list:
+      state = [state]
+    with self._graph._get_tf("Graph").as_default():
+      feed_dict = self._create_feed_dict(state)
+      probabilities = self._graph.session.run(self._pred_prob, feed_dict=feed_dict)
+      if deterministic:
+        return probabilities.argmax()
+      else:
+        return np.random.choice(
+            np.arange(self._env.n_actions), p=probabilities[0])
+
+  def restore(self):
+    """Reload the model parameters from the most recent checkpoint file."""
+    last_checkpoint = tf.train.latest_checkpoint(self._graph.model_dir)
+    if last_checkpoint is None:
+      raise ValueError('No checkpoint found')
+    with self._graph._get_tf("Graph").as_default():
+      variables = tf.get_collection(
+          tf.GraphKeys.GLOBAL_VARIABLES, scope='global')
+      saver = tf.train.Saver(variables)
+      saver.restore(self._graph.session, last_checkpoint)
+
+  def _create_feed_dict(self, state):
+    """Create a feed dict for use by predict() or select_action()."""
+    feed_dict = dict((f.out_tensor, np.expand_dims(s, axis=0))
+                     for f, s in zip(self._features, state))
+    return feed_dict
+
+  def _run_episodes(self, steps, temperature, saver, adapt_puct):
+    """Simulate the episodes for one iteration."""
+    buffer = []
+    self._env.reset()
+    root = TreeSearchNode(0.0)
+    for step in range(steps):
+      prob, reward = self._do_tree_search(root, temperature, adapt_puct)
+      state = self._env.state
+      if not self._state_is_list:
+        state = [state]
+      buffer.append((state, prob, reward))
+      action = np.random.choice(np.arange(self._env.n_actions), p=prob)
+      self._env.step(action)
+      if self._env.terminated:
+        self._env.reset()
+        root = TreeSearchNode(0.0)
+      else:
+#        root = root.children[action]
+        root = TreeSearchNode(0.0)
+      if time.time() > self._checkpoint_time:
+        self._checkpoint_index += 1
+        saver.save(
+            self._graph.session, self._graph.save_file, global_step=self._checkpoint_index)
+        self._checkpoint_time = time.time()
+    return buffer
+
+  def _optimize_policy(self, buffer, epochs):
+    """Optimize the policy based on the replay buffer from the current iteration."""
+    batch_size = self._graph.batch_size
+    n_batches = len(buffer) // batch_size
+    for epoch in range(epochs):
+      np.random.shuffle(buffer)
```

Comment:
What's the rationale for the shuffle? Note that this mutates a passed in argument (Though here it doesn't matter because the caller doesn't use buffer anymore).

Reply:
The same reason you would do a shuffle on any optimization: avoid artifacts by making the division into batches different for each epoch.